adapters/adkernelAdn/adkernelAdn.go

```diff
@@ -0,0 +1,254 @@
+func dispatchImpressions(imps *[]openrtb.Imp, impsExt *[]openrtb_ext.ExtImpAdkernelAdn) (map[openrtb_ext.ExtImpAdkernelAdn][]openrtb.Imp, *[]error) {
+	res := make(map[openrtb_ext.ExtImpAdkernelAdn][]openrtb.Imp)
+	errors := make([]error, 0)
+	for idx := range *imps {
+		imp := (*imps)[idx]
+		err := compatImpression(&imp)
+		if err != nil {
+			errors = append(errors, err)
+			continue
+		}
+		impExt := (*impsExt)[idx]
+		if res[impExt] == nil {
+			res[impExt] = make([]openrtb.Imp, 0)
+		}
+		res[impExt] = append(res[impExt], imp)
+
+	}
+	return res, &errors
+}
+
+//Alter impression info to comply with adkernel platform requirements
+func compatImpression(imp *openrtb.Imp) error {
+	imp.Ext = nil //do not forward ext to adkernel platform
+	if imp.Banner != nil {
+		return compatBanerImpression(imp.Banner)
+	}
+	return nil
+}
+
+func compatBanerImpression(banner *openrtb.Banner) error {
+	//As banner.w/h are required fields for adkernel adn platform - take the first format entry
+	if banner.W == nil && banner.H == nil {
+		if len(banner.Format) == 0 {
+			return newBadInputError(fmt.Sprintf("Expected at least one banner.format entry or explicit w/h"))
+		}
+		format := banner.Format[0]
+		banner.Format = banner.Format[1:]
+		banner.W = &format.W
+		banner.H = &format.H
+	}
+	return nil
+}
+
+func getImpressionExt(imp *openrtb.Imp) (*openrtb_ext.ExtImpAdkernelAdn, error) {
+	var bidderExt adapters.ExtImpBidder
+	if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
+		return nil, &errortypes.BadInput{
+			Message: err.Error(),
+		}
+	}
+	var adkernelAdnExt openrtb_ext.ExtImpAdkernelAdn
+	if err := json.Unmarshal(bidderExt.Bidder, &adkernelAdnExt); err != nil {
+		return nil, &errortypes.BadInput{
+			Message: err.Error(),
+		}
+	}
+	return &adkernelAdnExt, nil
+}
+
+func buildAdapterRequest(prebidBidRequest *openrtb.BidRequest, params *openrtb_ext.ExtImpAdkernelAdn, imps *[]openrtb.Imp) (*adapters.RequestData, error) {
+	newBidRequest := createBidRequest(prebidBidRequest, params, imps)
+	reqJSON, err := json.Marshal(newBidRequest)
+	if err != nil {
+		return nil, err
+	}
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+	headers.Add("x-openrtb-version", "2.5")
+
+	return &adapters.RequestData{
+		Method:  "POST",
+		Uri:     buildEndpointURL(params),
+		Body:    reqJSON,
+		Headers: headers}, nil
+}
+
+func createBidRequest(prebidBidRequest *openrtb.BidRequest, params *openrtb_ext.ExtImpAdkernelAdn, imps *[]openrtb.Imp) *openrtb.BidRequest {
+	bidRequest := *prebidBidRequest
+	bidRequest.Imp = *imps
+	for idx := range bidRequest.Imp {
+		imp := &bidRequest.Imp[idx]
+		imp.TagID = imp.ID
+	}
+	if bidRequest.Site != nil {
+		bidRequest.Site.Publisher = nil
+		bidRequest.Site.Domain = ""
+	}
+	if bidRequest.App != nil {
+		bidRequest.App.Publisher = nil
+	}
+	//nullify as prebid-server passes base64 adapter->uid map for all adapters
```

Comment:
er... how sure are you that you want to do this?

We didn't document t his feature well, so I can see how you got this impression... but actually the base64 adapter->uid map is just a fallback. Companies hosting PBS _should_ have their app config set so that PBS reads the user's ID from their domain's cookie.

Reply:
My mistake - I tested local PBS instance without properly configure `host_cookie.*`. With updated config there is no need to nullify this field.